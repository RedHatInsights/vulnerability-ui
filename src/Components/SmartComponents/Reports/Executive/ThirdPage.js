import React, { Fragment } from 'react';
import PropTypes from 'prop-types';
import {
    Section,
    Column,
    Chart,
    Table,
    Panel,
    PanelItem,
    InsightsLabel
} from '@redhat-cloud-services/frontend-components-pdf-generator';

import { Text } from '@react-pdf/renderer';
import { truncate } from 'lodash';

import styles from '../Common/styles';
import messages from '../../../../Messages';

const ThirdPage = ({ data, intl }) => {

    const rulesDesc = (rule, associatedCves) => (
        <Fragment>
            <Text style={styles.mb8}> {truncate(rule, { length: 300 })}</Text>
            <Text >
                {intl.formatMessage(messages.associatedCves)} {associatedCves.map(cves => cves).join(', ')}
            </Text>
        </Fragment>
    );

    const { topRules, rulesHeader, rulesBySeverityMap, rulesChart } = data;

    return (
        <Fragment key="first-section">
            <Text style={styles.header}>{intl.formatMessage(messages.rules)}</Text>
            <Section
                titleProps={{ style: styles.sectionTitle }}
                title={intl.formatMessage(messages.executiveReportRulesSection)}
            >
                <Column>
                    <Table
                        withHeader
                        rows={[
                            rulesHeader,
                            ...rulesBySeverityMap
                        ]}
                    />
                </Column>
                <Column>
                    <Chart
                        chartType="bar"
                        subTitle="rules"
                        title="Rules"
                        data={rulesChart}
                        colorSchema="orange"
                        legend={false}
                    />
                </Column>
            </Section>
            <Section
                titleProps={{ style: styles.sectionTitle }}
                title={intl.formatMessage(messages.executiveReportTop3Rules)} withColumn={false}
            >
                {
                    topRules.map((rule, key) => (
                        <Panel
                            key={key}
                            title={rule.name}
                            description={rulesDesc(rule.description, rule.associated_cves)}
                        >
                            <PanelItem title={intl.formatMessage(messages.executiveReportRulePanelTitle)}>
                                <InsightsLabel variant={rule.severity}/>
                            </PanelItem>
                            <PanelItem title={intl.formatMessage(messages.systemsExposed)}>
                                {rule.systems_affected.toString()}
                            </PanelItem>
                        </Panel>
                    ))

                }
            </Section>
        </Fragment>
    );
};

ThirdPage.propTypes = {
    data: PropTypes.exact({
        topRules: PropTypes.array,
        rulesHeader: PropTypes.array,
        rulesBySeverityMap: PropTypes.array,
        rulesChart: PropTypes.array
    }),
    intl: PropTypes.any
};

export default ThirdPage;
