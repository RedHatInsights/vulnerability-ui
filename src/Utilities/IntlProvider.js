import { createIntl, createIntlCache } from 'react-intl';

const cache = createIntlCache();
const locale = navigator.language.slice(0, 2);

export const intl = createIntl({
    locale
}, cache);

export const intlSettings = { locale };
