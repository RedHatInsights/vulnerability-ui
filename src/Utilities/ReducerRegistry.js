import promiseMiddleware from 'redux-promise-middleware';
import { getRegistry } from '@red-hat-insights/insights-frontend-components';

const registry = getRegistry({}, [promiseMiddleware()]);

export default {
    getStore: () => registry.getStore(),
    register: (newReducers) => {
        registry.register(newReducers);
    }
};
