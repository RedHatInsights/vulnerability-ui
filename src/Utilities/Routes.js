import React from 'react';
import some from 'lodash/some';
import PropTypes from 'prop-types';
import { Redirect, Route, Switch, useLocation } from 'react-router-dom';
import asyncComponent from './asyncComponent';
import { PATHS } from '../Helpers/constants';

const SystemsPage = asyncComponent(() =>
    import(
        /* webpackChunkName: "Systems" */ '../Components/SmartComponents/SystemsPage/SystemsPage'
    )
);

const LandingPage = asyncComponent(() =>
    import(
        /* webpackChunkName: "LandingPage" */ '../Components/SmartComponents/LandingPage/LandingPage'
    )
);

const Upgrade = asyncComponent(() =>
    import(
        /* webpackChunkName: "Upgrade" */ '../Components/PresentationalComponents/StaticPages/UpgradePage'
    )
);

const CVEDetailsPage = asyncComponent(() =>
    import(
        /* webpackChunkName: "CVEDetailsPage" */ '../Components/SmartComponents/CVEDetailsPage/CVEDetailsPage'
    )
);

const SystemDetailsPage = asyncComponent(() =>
    import(
        /* webpackChunkName: "SystemDetailsPage" */ '../Components/SmartComponents/SystemDetailsPage/SystemDetailsPage'
    )
);
const Reports = asyncComponent(() =>
    import(/* webpackChunkName: "Reports" */ '../Components/SmartComponents/Reports/ReportsPage')
);

const InsightsRoute = ({ component: Component, rootClass = 'vulnerability', ...rest }) => {
    const root = document.getElementById('root');
    root.removeAttribute('class');

    root.classList.add(`page__${rootClass}`, 'pf-l-page__main');
    root.classList.add(`page__${rootClass}`, 'pf-c-page__main');
    root.setAttribute('role', 'main');

    return <Route {...rest} component={Component} />;
};

InsightsRoute.propTypes = {
    component: PropTypes.func,
    rootClass: PropTypes.string
};

export const Routes = () => {
    let location = useLocation();
    let path = location.pathname;
    return (
        <Switch>
            <Redirect from={PATHS.cvepagePath.to} to={PATHS.systemDetailsPage.to} />
            <InsightsRoute exact path={PATHS.cveDetailsPage.to} component={CVEDetailsPage} />
            <InsightsRoute exact path={PATHS.systemDetailsPage.to}  component={SystemDetailsPage} />
            <InsightsRoute exact path={PATHS.cvesPage.to} component={LandingPage} />
            <InsightsRoute exact path={PATHS.systemsPage.to} component={SystemsPage} />
            <InsightsRoute exact path={PATHS.upgrade.to} component={Upgrade} />
            <InsightsRoute exact path={PATHS.reports.to} component={Reports} />
            <Route
                render={() => (some(PATHS, p => p.to === path) ? null : <Redirect to={PATHS.cvesPage.to} />)}
            />
        </Switch>
    );
};

