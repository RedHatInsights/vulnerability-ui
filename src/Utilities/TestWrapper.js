import propTypes from 'prop-types';
import { BrowserRouter as Router } from 'react-router-dom';
import { Provider } from 'react-redux';
import { IntlProvider } from '@redhat-cloud-services/frontend-components-translations';

const TestWrapper = ({ children, store }) => {
    return (
        <IntlProvider locale="en">
            { store ?
                <Provider store={store}>
                    <Router>
                        {children}
                    </Router>
                </Provider> : children
            }
        </IntlProvider>
    );
};

TestWrapper.propTypes = {
    children: propTypes.node,
    store: propTypes.object
};

export default TestWrapper;
