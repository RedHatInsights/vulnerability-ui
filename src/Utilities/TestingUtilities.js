import React from 'react';
import PropTypes from 'prop-types';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';
import { MemoryRouter, Routes, Route } from 'react-router-dom';
import { IntlProvider } from 'react-intl';

export const ComponentWithContext = ({
    renderOptions = {},
    children
}) => {
    const mockStore = configureStore();

    return (
        <IntlProvider locale="en">
            <Provider store={renderOptions?.store || mockStore()}>
                <MemoryRouter initialEntries={renderOptions?.initialEntries || ['/']}>
                    {
                        renderOptions?.componentPath ? (
                            <Routes>
                                <Route>
                                    {children}
                                </Route>
                            </Routes>
                        ) : children
                    }
                </MemoryRouter>
            </Provider>
        </IntlProvider>
    );
};

ComponentWithContext.propTypes = {
    children: PropTypes.element,
    renderOptions: PropTypes.object
};
