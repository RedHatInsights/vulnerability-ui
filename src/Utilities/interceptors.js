import {
    authInterceptor,
    interceptor401,
    responseDataInterceptor
} from '@redhat-cloud-services/frontend-components-utilities/interceptors';
import axios from 'axios';

const instance = axios.create();

export function errorInterceptor(error) {
    if (!axios.isCancel(error)) {
        const errObject = { ...error };

        errObject?.response?.status === 503 && window.setReadOnlyBannerVisible?.(true);

        if (errObject?.response?.data?.errors) {
            let data = errObject.response.data.errors[0];

            if (typeof data.detail === 'object') {
                data.detail = data.detail.msg;
            }

            throw data;
        }

        throw error;
    }
}

const readOnlyBannerInterceptor = response => {
    // if has updated property, then it was a successful write request
    response.hasOwnProperty('updated') && window.setReadOnlyBannerVisible?.(false);
    return response;
};

instance.interceptors.request.use(authInterceptor);
instance.interceptors.response.use(responseDataInterceptor);
instance.interceptors.response.use(null, interceptor401);
instance.interceptors.response.use(readOnlyBannerInterceptor, errorInterceptor);

export default instance;
