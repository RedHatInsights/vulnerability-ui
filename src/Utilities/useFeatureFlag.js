import { useFlag, useFlagsStatus } from '@unleash/proxy-client-react';

export default (flag) => {
    const { flagsReady } = useFlagsStatus();
    const isFlagEnabled = useFlag(flag);
    return flagsReady ? isFlagEnabled : false;
};
