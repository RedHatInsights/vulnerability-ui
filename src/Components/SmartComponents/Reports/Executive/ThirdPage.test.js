import ThirdPage from './ThirdPage';
import configureStore from 'redux-mock-store';
import { intl } from '../../../../Utilities/IntlProvider';
import TestWrapper from '../../../../Utilities/TestWrapper';
import { render } from '@testing-library/react';

const mockStore = configureStore([store => next => action => {}]);
let store = mockStore({ });

const data = {
    "topRules": [
        {
            "associated_cves": [
                "CVE-2018-12126",
                "CVE-2018-12127",
                "CVE-2018-12130",
                "CVE-2019-11091"
            ],
            "description": "Installed package versions are unable to assess this vulnerability.\n",
            "name": "\"MDS\": CPU side-channel",
            "rule_id": "CVE_2018_12130_cpu_kernel|CVE_2018_12130_CPU_KERNEL_NEED_UPDATE",
            "severity": 3,
            "systems_affected": 98
        },
        {
            "associated_cves": [
                "CVE-2019-1125"
            ],
            "description": "Vulnerable package versions are installed.\n",
            "name": "\"SWAPGS\": CPU side-channel",
            "rule_id": "CVE_2019_1125_kernel|CVE_2019_1125_KERNEL_2",
            "severity": 3,
            "systems_affected": 98
        },
        {
            "associated_cves": [
                "CVE-2018-3620",
                "CVE-2018-3646"
            ],
            "description": "Installed package versions are unable to assess this vulnerability.\n",
            "name": "\"L1TF\": CPU side-channel with outdated kernel",
            "rule_id": "CVE_2018_3620_cpu_kernel|CVE_2018_3620_CPU_KERNEL_NEED_UPDATE",
            "severity": 3,
            "systems_affected": 96
        }
    ],
    "rulesHeader": [
        "Severity",
        "Num. security rules",
        "Num. affected systems"
    ],
    "rulesBySeverityMap": [
        [
            "Critical",
            "0",
            "0"
        ],
        [
            "Important",
            "11",
            "135"
        ],
        [
            "Moderate",
            "0",
            "0"
        ],
        [
            "Low",
            "0",
            "0"
        ]
    ],
    "rulesChart": [
        {
            "name": "Critical",
            "x": "Critical",
            "y": 0
        },
        {
            "name": "Important",
            "x": "Important",
            "y": 135
        },
        {
            "name": "Moderate",
            "x": "Moderate",
            "y": 0
        },
        {
            "name": "Low",
            "x": "Low",
            "y": 0
        }
    ]
}

describe('ThirdPage component', () => {
    it.skip('Should match snapshot', () => {
        const { asFragment } = render(
            <TestWrapper store={ store }>
                <ThirdPage data={data} intl={intl} />
            </TestWrapper>
        );

        expect(asFragment()).toMatchSnapshot();
    })
});
