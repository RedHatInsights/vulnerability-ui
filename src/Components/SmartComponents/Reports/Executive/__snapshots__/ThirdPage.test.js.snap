// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`ThirdPage component Should match snapshot 1`] = `
<Provider
  store={
    {
      "clearActions": [Function],
      "dispatch": [Function],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
>
  <BrowserRouter>
    <Router
      location={
        {
          "hash": "",
          "key": "default",
          "pathname": "/",
          "search": "",
          "state": null,
        }
      }
      navigationType="POP"
      navigator={
        {
          "action": "POP",
          "createHref": [Function],
          "createURL": [Function],
          "encodeLocation": [Function],
          "go": [Function],
          "listen": [Function],
          "location": {
            "hash": "",
            "key": "default",
            "pathname": "/",
            "search": "",
            "state": null,
          },
          "push": [Function],
          "replace": [Function],
        }
      }
    >
      <ThirdPage
        data={
          {
            "rulesBySeverityMap": [
              [
                "Critical",
                "0",
                "0",
              ],
              [
                "Important",
                "11",
                "135",
              ],
              [
                "Moderate",
                "0",
                "0",
              ],
              [
                "Low",
                "0",
                "0",
              ],
            ],
            "rulesChart": [
              {
                "name": "Critical",
                "x": "Critical",
                "y": 0,
              },
              {
                "name": "Important",
                "x": "Important",
                "y": 135,
              },
              {
                "name": "Moderate",
                "x": "Moderate",
                "y": 0,
              },
              {
                "name": "Low",
                "x": "Low",
                "y": 0,
              },
            ],
            "rulesHeader": [
              "Severity",
              "Num. security rules",
              "Num. affected systems",
            ],
            "topRules": [
              {
                "associated_cves": [
                  "CVE-2018-12126",
                  "CVE-2018-12127",
                  "CVE-2018-12130",
                  "CVE-2019-11091",
                ],
                "description": "Installed package versions are unable to assess this vulnerability.
",
                "name": ""MDS": CPU side-channel",
                "rule_id": "CVE_2018_12130_cpu_kernel|CVE_2018_12130_CPU_KERNEL_NEED_UPDATE",
                "severity": 3,
                "systems_affected": 98,
              },
              {
                "associated_cves": [
                  "CVE-2019-1125",
                ],
                "description": "Vulnerable package versions are installed.
",
                "name": ""SWAPGS": CPU side-channel",
                "rule_id": "CVE_2019_1125_kernel|CVE_2019_1125_KERNEL_2",
                "severity": 3,
                "systems_affected": 98,
              },
              {
                "associated_cves": [
                  "CVE-2018-3620",
                  "CVE-2018-3646",
                ],
                "description": "Installed package versions are unable to assess this vulnerability.
",
                "name": ""L1TF": CPU side-channel with outdated kernel",
                "rule_id": "CVE_2018_3620_cpu_kernel|CVE_2018_3620_CPU_KERNEL_NEED_UPDATE",
                "severity": 3,
                "systems_affected": 96,
              },
            ],
          }
        }
        intl={
          {
            "$t": [Function],
            "defaultFormats": {},
            "defaultLocale": "en",
            "defaultRichTextElements": undefined,
            "fallbackOnEmptyString": true,
            "formatDate": [Function],
            "formatDateTimeRange": [Function],
            "formatDateToParts": [Function],
            "formatDisplayName": [Function],
            "formatList": [Function],
            "formatListToParts": [Function],
            "formatMessage": [Function],
            "formatNumber": [Function],
            "formatNumberToParts": [Function],
            "formatPlural": [Function],
            "formatRelativeTime": [Function],
            "formatTime": [Function],
            "formatTimeToParts": [Function],
            "formats": {},
            "formatters": {
              "getDateTimeFormat": [Function],
              "getDisplayNames": [Function],
              "getListFormat": [Function],
              "getMessageFormat": [Function],
              "getNumberFormat": [Function],
              "getPluralRules": [Function],
              "getRelativeTimeFormat": [Function],
            },
            "locale": "en",
            "messages": {},
            "onError": [Function],
            "onWarn": [Function],
            "textComponent": Symbol(react.fragment),
            "timeZone": undefined,
          }
        }
      >
        <TEXT
          style={
            {
              "color": "#C9190B",
              "fontSize": 20,
            }
          }
        >
          Insights Security Rules
        </TEXT>
        <Section
          title="Security rules affecting systems"
          titleProps={
            {
              "style": {
                "fontSize": 10,
              },
            }
          }
        >
          <div />
        </Section>
        <Section
          title="Top 3 security rules in your infrastructure"
          titleProps={
            {
              "style": {
                "fontSize": 10,
              },
            }
          }
          withColumn={false}
        >
          <div />
        </Section>
      </ThirdPage>
    </Router>
  </BrowserRouter>
</Provider>
`;
